### ISSUE TITLE
NIP-65 minor change for better DM privacy?

### ISSUE THREAD

**Issue by fr4nzap** [nostr-protocol/nips#840]
I was reading NIP-65 and noticed

> DMs SHOULD only be broadcasted to the author's WRITE relays and to the receiver's READ relays to keep maximum privacy.

Since most users add lots of READ relays not to miss out on public notes, how about adding a "DM" marker to NIP-65 such that clients know exactly where to broadcast DMs to?  Privacy-sensitive users can now choose to add self-hosted relays (or reputable, dedicated DM relays, which might excel at spam filtering, require NIP-42 auth, etc) exclusively for DMs, and use clients that honor the `dm` marker.

This means not leaking metadata to random relays and would improve DM privacy more than what the NIP is currently stating.

The change would be backwards compatible and shifts the burden to users and clients rather than bloating the spec.

**Comment by staab** [comment]
I think this is a good idea, it's been suggested before, and might be relevant to https://github.com/nostr-protocol/nips/issues/790

**Comment by fiatjaf** [comment]
Yes, I agree.

**Comment by vitorpamplona** [comment]
Complete support from my side. 

I have a further need to identify relays by role: 
- A "public_mentions" relay, which others can send events and also can use to download events that cite the user. 
- A "private_mentions" relay, which others can send events but CANNOT use to download events that cite the user. 
- A "private_backup" relay to receive all new public and private events from the author BUT only the author can download. 
- A "public_outbox" relay that receives ONLY the public events from the author and allows anyone out there to download. 
- An "archive" relay to be used by others to download old events by the author or versions or replaceable events.

**Comment by vitorpamplona** [comment]
names can be refined, of course :)

**Comment by vitorpamplona** [comment]
The "private_backup" one doesn't have any utility for other users. Since NIP-65 is designed to advertise relays for others to use, this NIP not be the place for it.

**Comment by staab** [comment]
> The "private_backup" one doesn't have any utility for other users. Since NIP-65 is designed to advertise relays for others to use, this NIP not be the place for it.

This is true, but we do need an interoperable place to store those settings, which we don't currently have (kind 3 doesn't count). I think it would make sense to add them to 65 rather than create a new nip/kind.

**Comment by mikedilger** [comment]
How about we use an unsigned integer to indicate the usage, e.g.

bit 0 - Outbox
bit 1 - Inbox
bit 2 - DM (use for sending direct messages)
bit 3 - Open (author-tagged events written are explicitly publicly readable, except perhaps DMs)
bit 4 - Archive (author-created events are expected to be made available for a long period of time)

In this scheme, the tag name goes back to something simple

````json
"tags": [
  ["relay", "wss://...", "17"],
  ["relay", "wss://...", "10"],
]
````

**Comment by fiatjaf** [comment]
Part of the original idea of NIP-65 was that it would be very small that it wouldn't hurt to store it in every relay you could find under the sun, so it would be easy to find people's relays.

For private categories we should probably use NIP-51 relay sets: https://github.com/nostr-protocol/nips/pull/786/files

**Comment by arthurfranca** [comment]
Imo the categories for NIP-65 are getting out of control. It should be either in or out and the rest you identify with optional
extra attributes we may introduce with time, starting with "kinds" and "nips".
`["inbox", "wss://abc.xyz", "{ \"kinds\": [4, 14], \"nips\": [102] }"]` <- prefer this one for these kinds and nips

**Comment by fiatjaf** [comment]
`"{\"kinds\": [4, 14], \"nips\": [102] }"` is too complex, too broad. It's unimplementable, it's asking to not be used.

**Comment by vitorpamplona** [comment]
> This is true, but we do need an interoperable place to store those settings, which we don't currently have (kind 3 doesn't count). I think it would make sense to add them to 65 rather than create a new nip/kind.

There are two choices here: 
1. We could encrypt relay settings in the content of `kind:10002` in a similar way with do with private lists. 
2. We can create a copycat event (`kind:10003`?) to store a relay list for the user's own clients and no one else. 

Option 1 will make this event larger, but because it is private, it clearly differentiates between my relay list and the relay list I want people to use when downloading my events. 

Option 2 allows for public and private relay lists my the user's personal stuff and keeps 10002 as small as possible (I am not sure if size is an actual issue)

**Comment by frnandu** [comment]
I was under the impression that private/public relays list would be added to nip-51 by https://github.com/nostr-protocol/nips/pull/786

**Comment by vitorpamplona** [comment]
> I was under the impression that private/public relays list would be added to nip-51 by #786

This works for me as well. 3 options then. Would it follow the same makers as here or would it just be a list of `r` tags?

**Comment by sondreb** [comment]
**tldr; REMOVE READ/WRITE, 10002 should be list of user relays (replacing kind 3 content), publish 10002 when you reply on new relays. That's it.**

It is enough to have the READ ("INBOX") and WRITE ("OUTBOX") flags, don't need additional types.

What instead should be done, is change some of the NIP-65, especially this sentence:

```DMs SHOULD only be broadcasted to the author's WRITE relays and to the receiver's READ relays to keep maximum privacy.```

My suggestion is to rephrase this to:

```DMs SHOULD only be broadcasted to READ relays, including the READ relays of the author. Not publishing the DMs to the WRITE, increases privacy and it keeps all relevant DMs accessible on the same (READ) relay.```

This also needs one additional bullet point:

```
When broadcasting an event, Clients SHOULD:

Broadcast the event to the WRITE relays of the author
Broadcast the event to all READ relays of each tagged user
```

I suggest changing to:

```
When broadcasting an event, Clients SHOULD:

Broadcast the event to the WRITE relays of the author
Broadcast the event to the READ relays of the author
Broadcast the event to all READ relays of each tagged user
```

Why you ask?

Because READ relays are for DMs and replies. It's a single place where I as a user, can go and get all my DMs (incoming and outgoing) and all my replies (incoming and outgoing).

It makes no sense to separate incoming DMs and outgoing DMs. It means two different relays must be connected to, events must be gotten from both of them and then threaded together. What if one relays is a bit slow that day? What should the UI do then, in a thread it will not look good and not be a good UX.

If my own published events are not fully available on my READ relays, then this whole NIP loses a lot of the usefulness. People will just use 3-4 relays and make them all WRITE/READ.

With my suggested changes, I get to have a fully contained relay where I can easily get my engagement content, I also can have a relay that people can connect to when they open my profile to view my public posts (my WRITE relay).

The way the specification is written no, it makes very little sense to me. I thought I implemented NIP-65 years ago, but I completely misunderstood it. And now after many hours of discussing it, reading it, looking at code, I'm even more confused and I think this is very much against the ethos of Nostr.

Clients should then use the users READ relays to get all the events relevant for the user, the original posts, all their replies to that post, all their outgoing and incoming DMs.

...

Now here is the catch: The way this NIP is designed today, clients are supposed to connect to WRITE relay to get public posts. But then if I want to look at engagement (replies), I must actually connect to the user's READ relays. Suddenly I need to have a lot more relay connections just for this single user.

This makes me think that the whole NIP should be redone into something simpler:

**kind 10002 for having a short and simple list of relays, replacing content on kind 3.
Use this list, when you want to publish replies and DMs to someone, or you want to look at their public content.
When you reply to someone and your own kind 10002 is not on the relay, publish it together with your reply, so that everyone can see where to look you up.**

That's it, all that was needed.

All it should do is have a list of relays, nothing more, no types. It adds complexity without any real value.

I'm sorry if I sound a bit frustrate here, but this feels like an extreme waste of many hours trying to understand this, by just finally realizing the whole concept is flawed and could be simplified a whole lot.

**Comment by mikedilger** [comment]
First, new NIP-17 DMs don't use NIP-65 they go on DM relays advertised in a kind 10050 (separate).  Reasoning is that separate apps built on nostr should be isolated from each other and not all mixed together.  I actually wasn't very involved in that decision, I just accepted it.  But old NIP-04 DMs go into your inbox.

As for inboxes (read) and outboxes (write), the split was done because some people wanted different rulesets based on the purpose.  For an outbox relay, I want everybody to see those notes, the whole world should be able to read my important and funny posts without authenticating.  But for an inbox relay, sometimes I would prefer privacy.  And maybe I want to prevent spam.  If someone sends me a note to my inbox I might want them to authenticate, and I might not want that to be world viewable (If the sender does want it world viewable, they can copy it to their own outbox).

Outbox means "I'm making this public".  Inbox means "I'm receiving events addressed to me here."

Yes most people just make all their relays both inbox and outbox.  That's fine.  But the separation is useful for some people with different ways of working.  I'm one of those people with different rules on my inbox.

When you send your notes to the world on your outbox relay, you are free to also post them to your inbox relays if you want the optimization you mentioned -- this way you can read all the notes from your inboxes -- both your notes and other people's notes -- without connecting to your outboxes at all.  BUT we didn't specify that anywhere and clients aren't doing that.  I'm not sure it is worth it to make such a change, but I'll let others weigh in.

**Comment by sondreb** [comment]
What is happening to Nostr? Having a separate kind for DM relays? This is utterly fragmenting the individual user's data. This will not scale in terms of adoption, headaches for users and developers. The UX for this will be nuts.

What is this reasoning for building an utterly fragmented environment where a person's data is scattered across multiple different servers, multiple different protocols? That is not what decentralization is about, it's just making a mess for everyone, the user and developers.

Decentralization is about separating each user, so we can all scale individually. If I have all my content scattered everywhere, we've just replaced the existing world wide web like X, YouTube, Messenger, etc.

```Reasoning is that separate apps built on nostr should be isolated from each other and not all mixed together.```

Why cannot a pure messenger app use relays from kind:10002? Data is already separated with kinds and other tags on events, physical separation brings little of value and more complexity.

I will implement NIP-17 soon and I will ignore kind:10050. Relays simply should only relay DM events based upon auth, that solves a lot of things.

```If someone sends me a note to my inbox I might want them to authenticate, and I might not want that to be world viewable (If the sender does want it world viewable, they can copy it to their own outbox).```

Private replies to public threads? Copy the OP and send the person an e-mail using SMTP.

```When you send your notes to the world on your outbox relay, you are free to also post them to your inbox relays if you want the optimization you mentioned -- this way you can read all the notes from your inboxes -- both your notes and other people's notes -- without connecting to your outboxes at all. BUT we didn't specify that anywhere and clients aren't doing that. I'm not sure it is worth it to make such a change, but I'll let others weigh in.```

Better clarification in the NIP would help, I will at least do what you suggest here, ignore the READ/WRITE.

Sorry for the tone of my replies here, I'm just baffled about how the growth of complexity has made Nostr something it did not use to be.

**Comment by mikedilger** [comment]
If all of a user's metadata were stuffed into the same event (kind 0) including relays, including new DM relays, cashu stuff, etc.... then we have many systems writing that same event, with many more opportunities for one system to fuck up the event and the damage hurts all the other systems.  Isolating the applications running on top of nostr by giving them all their own event kinds is beneficial.  Each application can do things differently from each other application.  Clients can choose any subset of applications to support, and don't get confused by other applications stuff because it is all disjoint.

That being said I think of some things as fundamental to the layer BELOW the applications.  Relay choice by kind 10002 is used by many applications. Any kind of key rollover is also at the lower layer.

But Kind-17 DMs are a new application.  They simply decided to use their own application-layer relay choice mechanism.  It doesn't break anything, it's not obviously better or worse IMHO.  I coded it, it wasn't a mess, and the UX is fine.

**Comment by vitorpamplona** [comment]
NIP-17 DMs have their own relay list kind
